### PR TITLE
Support custom prediction functions in `SklearnModelWrapper`

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/11577?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11577/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11577
```

</p>
</details>

### Related Issues/PRs

#11559

### What changes are proposed in this pull request?

The `predict_log_proba` method is one of prediction methods of [LogisticRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html) model, but we accidentally dropped support for it when `_SklearnModelWrapper` was introduced in MLflow 2.6.0. It works with prior versions of MLflow.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

```
import mlflow 
from mlflow.models.signature import infer_signature
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import load_iris 

# load the iris dataset
iris = load_iris(as_frame=True)
x = iris.data
y = iris.target

# infer signature
signature = infer_signature(model_input=x, model_output=y)
input_example = x.iloc[0:10]

with mlflow.start_run(run_name="logging_with_custom_predict_function") as run:

    rfc = RandomForestClassifier()
    rfc.fit(x, y)

    # Log the model
    mlflow.sklearn.log_model(
        sk_model=rfc,
        artifact_path = rfc.__class__.__name__,
        signature=signature,
        input_example=input_example,
        pyfunc_predict_fn = "predict_log_proba"
        )
    
# load model using pyfunc
model_uri = f"runs:/{run.info.run_id}/{rfc.__class__.__name__}"
model = mlflow.pyfunc.load_model(model_uri)

# predict
model.predict(x)
```

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Fix regression of scikit-learn model flavor, where the `predict_log_proba` and `predict_joint_log_probe` method couldn't be specified as pyfunc predict function via `pyfunc_predict_fn` parameter.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/sklearn/__init__.py [hunk 1]
```diff
@@ -534,6 +534,10 @@ def predict_proba(self, *args, **kwargs):
         if hasattr(self.sklearn_model, "predict_proba"):
             return self.sklearn_model.predict_proba(*args, **kwargs)
 
+    def predict_log_proba(self, *args, **kwargs):
```
Comment (harupy) [inline mlflow/sklearn/__init__.py:537]: It looks like there is one more predict variant: https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.GaussianNB.html#sklearn.naive_bayes.GaussianNB.predict_joint_log_proba

**Review by harupy — APPROVED** [review]
https://github.com/mlflow/mlflow/pull/11577/files#r1547063741 is not a blocker. LGTM!

File: mlflow/sklearn/__init__.py (hunk 1 above)
Comment (B-Step62) [reply mlflow/sklearn/__init__.py:537]: Oh great catch! This one has exactly same signature so adding that should be fine. In the future we may want to implement better delegation mechanism.

File: tests/sklearn/test_sklearn_model_export.py [hunk 2]
```diff
@@ -768,12 +777,18 @@ def test_log_model_with_code_paths(sklearn_knn_model):
...
 
-def test_log_predict_proba(sklearn_logreg_model):
-    model, inference_dataframe = sklearn_logreg_model
-    expected_scores = model.predict_proba(inference_dataframe)
+@pytest.mark.parametrize(
+    "predict_fn", ["predict", "predict_proba", "predict_log_proba", "predict_joint_log_proba"]
+)
+def test_log_model_with_custom_pyfunc_predict_fn(sklearn_gaussian_model, predict_fn):
+    if Version(sklearn.__version__) < Version("1.2.0") and predict_fn == "predict_joint_log_proba":
+        pytest.skip("predict_joint_log_proba is not available in scikit-learn < 1.2.0")
```
Comment (harupy) [inline tests/sklearn/test_sklearn_model_export.py:785]: /ai teach me a better way to write this code using `pytest.param`.

Comment (B-Step62) [reply tests/sklearn/test_sklearn_model_export.py:785]: hmm I think this is less clear...?

Comment (harupy) [reply tests/sklearn/test_sklearn_model_export.py:785]: Not a blocker, I'd use `pytest.param` in this case to simplify the test function body.

Comment (B-Step62) [reply tests/sklearn/test_sklearn_model_export.py:785]: Imo adding ^ block to remove simple 2 lines if-statement doesn't pay off....

**Pull request merged**